pkg/threescale/backend/manager.go

```diff
@@ -256,48 +259,48 @@ func (b *Backend) computeAffectedMetrics(hierarchy api.Hierarchy, metrics api.Me
 	return metrics
 }
 
-// isAuthorized works on a copy of the passed Application and will inform the caller if limits have
-// been breached in the provided affected metrics
-func (b *Backend) isAuthorized(application Application, affectedMetrics api.Metrics) (Application, bool) {
-	copiedApplication := application.deepCopy()
+// isAuthorized takes a read lock on the application and confirms if the request
+// should be authorized based on the affected metrics against current state
+func (b *Backend) isAuthorized(application Application, affectedMetrics api.Metrics) bool {
+	application.RLock()
+	defer application.RUnlock()
+
 	commitChanges := true
 
 out:
 	for metric, incrementBy := range affectedMetrics {
-		cachedValue, ok := copiedApplication.LimitCounter[metric]
+		cachedValue, ok := application.LimitCounter[metric]
 		if ok {
 			for _, limit := range cachedValue {
-				newValue := limit.current + incrementBy
-				if newValue > limit.max {
+				if limit.current+incrementBy > limit.max {
 					commitChanges = false
 					break out
 				}
-				limit.current = newValue
 			}
-		} else {
-			copiedApplication = b.updateUnlimitedCounter(metric, incrementBy, copiedApplication)
 		}
 	}
-	return copiedApplication, commitChanges
+	return commitChanges
 }
 
-// localReport reports to the cache
-func (b *Backend) localReport(cacheKey string, application Application, affectedMetrics api.Metrics) Application {
-	copiedApplication := application.deepCopy()
+// localReport takes a write lock on the application and reports to the cache
+func (b *Backend) localReport(cacheKey string, metrics api.Metrics) {
+	application, _ := b.cache.Get(cacheKey)
```

Comment:
can't miss?

Reply:
can miss but returns an empty `Application` if it does, however you have made me check the code and I see that the maps should be initialised if we miss so refactoring this.